NaI implementation and ROOT output

spandeyehep opened the pull request:
New files:

k100_ROOTOut.* : Produces output file in ROOT tree formatk100_NaIParametrization.* : for NaI array detector placement
k100_nrCascadeTree.* : Reads inputs from nrCascadeSim file
k100_SteppingAction.* : Does nothing at the moment but can be used for debugging
spandey_nCap.mac : Macro for PuBe source (placed outside the shielding) with NaIArray
spandey_nCap_cylinder_source.mac : Macro for thermal neutron source in the silicon crystal
neutron_capture.dat : energy spectrum for thermal neutrons


Changes:
k100Sim.cc : Updated to accept nrCascadeSim file name as input argument (if required)
k100_PrimaryGeneratorAction.* : Added code snippet to read nrCascadeSim file and generate corresponding photons
k100_DetectorConstruction.* : Implement NaI array
k100_DetectorConstructionMessenger.* : Activate/deactivate NaI array using macro command (/CDMS/detector/activate NaIArray)
k100_DataStorage.* : Store output in ROOT tree format
k100_ZipSD.* : Using this code base to include NaI simHits in the output collection

CMakeLists.txt : Extended cmake version dependence to run on cdms2 machine

villaa commented:
This is great, let's make the target branch develop though--this way we can test and then release a new version number.

villaa commented:
Also, open up an issue detailing the enhancements requested--that this PR will solve.

spandeyehep commented:
I have changed the base branch to "dev". Does it look okay now? If yes, then I will create an issue documenting the enhancements.

villaa commented:
As mentioned in meeting: before merging this we have to restore the text file output @spandeyehep let me know when that's done.

spandeyehep commented:
I have added a flag (`-t`) in the code to switch from root output format to text format.
`./k00sim <macro.mac>` : produces ROOT file.
`./k00sim -t <macro.mac>`: produces text file.
@villaa

villaa commented:
@spandeyehep good update! I would like the text file to be the default behavior so that it is compatible with previous runs. So please change it so that the `-R` flag (or similar) writes to a root file instead.

spandeyehep commented:
Hi @villaa . Done!!
I have removed `-t` flag, and added `-r` flag. The behavior is now swapped.  
`./k00sim <macro.mac>` : produces text file.
`./k00sim -r <macro.mac>`: produces ROOT file.

villaa commented:
@spandeyehep I did a review of the code changes and there are some things I don't understand; you should be able to see the comments. 

The key thing I don't understand is in PrimaryGeneratorAction--it seems to me that it _only_ allows running where nrCascadeTree is supplied, but what about just the PuBe source with gammas?

spandeyehep commented:
@villaa where can I find the comments? I don't see it.

The code requires nrCascadeTree file only when `-cutomgen` flag is supplied [1]-[2] , otherwise the behavior is default, i.e. use source provided in macro as input and do PuBe gammas if it has been set to `True` as macro-command [3].

[1] https://github.com/spandeyehep/k100Sim-1/blob/k100sim_NaIArray/k100Sim.cc#L197-L200
[2] https://github.com/spandeyehep/k100Sim-1/blob/k100sim_NaIArray/src/k100_PrimaryGeneratorAction.cc#L103-L158
[3] https://github.com/spandeyehep/k100Sim-1/blob/k100sim_NaIArray/src/k100_PrimaryGeneratorAction.cc#L159-L231

villaa commented:
> @villaa where can I find the comments? I don't see it.
> 
> The code requires nrCascadeTree file only when `-cutomgen` flag is supplied [1]-[2] , otherwise the behavior is default, i.e. use source provided in macro as input and do PuBe gammas if it has been set to `True` as macro-command [3].
> 
> [1] https://github.com/spandeyehep/k100Sim-1/blob/k100sim_NaIArray/k100Sim.cc#L197-L200 [2] https://github.com/spandeyehep/k100Sim-1/blob/k100sim_NaIArray/src/k100_PrimaryGeneratorAction.cc#L103-L158 [3] https://github.com/spandeyehep/k100Sim-1/blob/k100sim_NaIArray/src/k100_PrimaryGeneratorAction.cc#L159-L231

So, `sourceGun==True` means we are using a capture file?

villaa commented:
@spandeyehep I think I see now--what was confusing me is that I had the `sourceGun` variable before; but it did something useless: threw isotropic 8 MeV neutron. So you just removed that functionality and made it the nrCascadeTree input. 

That's fine, but I might suggest changing the name of `sourceGun` and `useGun` to something more obvious, like `throwCaptures` or something...

spandeyehep commented:
@villaa Sure, I will update the flag name to make it less confusing.
I will probably push the changes once I have also implemented and validated boron shield. Let me know if that sounds good to you.

spandeyehep commented:
Hi @villaa , unfortunately I couldn't get much time to work on boron shield. I hope to implement that in a couple of days.
I agree with your recommendation. You can go ahead and merge the PR. I will open another request once boron shield is implemented.

villaa commented:
> Hi @villaa , unfortunately I couldn't get much time to work on boron shield. I hope to implement that in a couple of days. I agree with your recommendation. You can go ahead and merge the PR. I will open another request once boron shield is implemented.

@spandeyehep did you update the names of `sourceGun` and `useGun` as per above? That I would like to do before merging this PR.

spandeyehep commented:
@villaa I have updated the variable names `sourceGun` and `useGun` to `throwCaptures` and `useCapture`, resp.

villaa reviewed (approved):
Thanks for updating the variable names for  `throwCaptures` etc.

spandeyehep commented on src/k100_SteppingAction.cc:
```diff
@@ -0,0 +1,48 @@
+{}
+
+//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......                                                                                                                                
+
```
yes, indeed. I suggest lets keep it as it is in case one wants to check states of G4Steps in the passive geometry. I can remove it if you have any objection.

spandeyehep commented on src/k100_DataStorage.cc:
```diff
@@ -60,21 +61,39 @@ k100_DataStorage::k100_DataStorage(G4String filename, G4int run, G4int rseed)
     VariableNames[kk] = myNames[kk];
 
 
-#ifdef ASCIIOUT
```
Agreed. Will clean up in the next iteration.

spandeyehep commented on include/k100_DataStorage.hh:
```diff
@@ -6,15 +6,15 @@
        
 
```
Yes. That is correct. I wanted to keep compiler directives as it is but then I was having trouble to switch it to text or root format after compilation.  I couldn't find a better way to conditionally select a particular compiler directive.
